### PR TITLE
Fix `EncodedFile.writelines`

### PR DESCRIPTION
**Pull request by blueyed**
This is implemented by IOBase already, which additionally checks if the
stream is not closed, and calls `write` per line.

Ref/via: https://github.com/pytest-dev/pytest/pull/6558#issuecomment-578210807

TODO:

- [x] test: could check if it calls through to `self._checkClosed()` (https://github.com/cpython/cpython/blob/58f525a216e651f1e17a18d31f190db727ed3de5/Lib/_pyio.py#L589) now
- [x] changelog

### CODE REVIEW

**Comment by blueyed** [comment]
Oh.. it is not even covered.

I've found that it appears to be broken currently even, since it assumes a list of strs, while it should be bytes: https://docs.python.org/3/library/asyncio-stream.html?highlight=writelines#asyncio.StreamWriter.writelines

**Comment by blueyed** [comment]
Rebased on master since it appears to be a bugfix.

**Comment by blueyed** [comment]
/cc @gavento 
Does this look good to you?

**Comment by blueyed** [comment]
@bluetech please check if the typing things are proper here.

File: src/_pytest/capture.py [hunk 1]
```diff
@@ -413,30 +415,27 @@ def safe_text_dupfile(f, mode, default_encoding="UTF8"):
...
             raise TypeError(
                 "write() argument must be str, not {}".format(type(obj).__name__)
             )
-        return self.buffer.write(obj)
+        return self.buffer.write(obj.encode(self.encoding, "replace"))
 
-    def writelines(self, linelist):
-        data = "".join(linelist)
-        self.write(data)
+    def writelines(self, linelist: List[str]) -> None:
```
Comment (bluetech) [inline src/_pytest/capture.py]: typeshed uses Iterable here: https://github.com/python/typeshed/blob/d5851eca6f448b648e1a7b3f3010cc3dde498535/stdlib/3/typing.pyi#L490

Would also change the argument name to just `lines`.

File: src/_pytest/capture.py [hunk 2]
```diff
@@ -413,30 +415,27 @@ def safe_text_dupfile(f, mode, default_encoding="UTF8"):
...
+        self.buffer.writelines([x.encode(self.encoding, "replace") for x in linelist])
```
Comment (bluetech) [inline src/_pytest/capture.py]: Any reason to use a list comprehension and not generator expression (omit the `[]`)?

File: testing/test_capture.py [hunk 3]
```diff
@@ -831,7 +833,7 @@ def test_dontreadfrominput():
 
 
 @pytest.fixture
-def tmpfile(testdir):
+def tmpfile(testdir) -> Generator[BinaryIO, None, None]:
```
Comment (bluetech) [inline testing/test_capture.py]: Generally better to use `Iterator[X]` over `Generator[X, None, None]` - shorter and stricter.

Hmm but now that I think about it, `pytest.fixture` actually would want a real generator, I think? If so, then yeah, better keep it.

File: src/_pytest/capture.py [hunk 4]
```diff
@@ -413,30 +415,27 @@ def safe_text_dupfile(f, mode, default_encoding="UTF8"):
 class EncodedFile:
     errors = "strict"  # possibly needed by py3 code (issue555)
```
Comment (bluetech) [inline src/_pytest/capture.py]: A bit strange that it says `strict` here but actually goes to use `replace`.

File: src/_pytest/capture.py [hunk 5]
```diff
@@ -413,30 +415,27 @@ def safe_text_dupfile(f, mode, default_encoding="UTF8"):
...
-    def __init__(self, buffer, encoding):
+    def __init__(self, buffer: BinaryIO, encoding: str) -> None:
         self.buffer = buffer
         self.encoding = encoding
 
-    def write(self, obj):
-        if isinstance(obj, str):
-            obj = obj.encode(self.encoding, "replace")
-        else:
+    def write(self, obj: str) -> int:
```
Comment (bluetech) [inline src/_pytest/capture.py]: Would change the argument name `obj` to `s` while at it, `obj` is a somewhat misleading name for it.

**Comment by bluetech** [comment]
If we're looking for potential breakage, previously `EncodedFile` only required the buffer to implement `write`, now it also requires `writelines`. I didn't examine how `EncodedFile` is used, so this might be fine.

File: src/_pytest/capture.py (hunk 2 above)
Comment (blueyed) [reply src/_pytest/capture.py]: I wanted to please the typing (which was `List`).

File: src/_pytest/capture.py (hunk 4 above)
Comment (blueyed) [reply src/_pytest/capture.py]: Yeah, likely an oversight.  Ref to issue: https://bitbucket.org/pytest-dev/pytest/issues/555
Will create a followup issue.

**Comment by blueyed** [comment]
> If we're looking for potential breakage, previously `EncodedFile` only required the buffer to implement `write`, now it also requires `writelines`. I didn't examine how `EncodedFile` is used, so this might be fine.

Will check this / add a test for this in the next iteration - could likely fall back to the old behavior of using `write` then.

**Comment by blueyed** [comment]
> > If we're looking for potential breakage, previously `EncodedFile` only required the buffer to implement `write`, now it also requires `writelines`. I didn't examine how `EncodedFile` is used, so this might be fine.
> 
> Will check this / add a test for this in the next iteration - could likely fall back to the old behavior of using `write` then.

Should be safe, since we only use `EncodedFile` via `safe_text_dupfile` for the `tmpfile` in `FDCaptureBinary`.

**Comment by blueyed** [comment]
Waiting for another approval / re-review.

**Comment by nicoddemus** [comment]
Bummer, a "fixup!" commit bf5c76359cf9dce7a94989d7f9edd4e22e6ffa3a got merged.

**Pull request merged**